geode-core/src/main/java/org/apache/geode/cache/client/internal/QueueManagerImpl.java

```diff
@@ -926,6 +930,9 @@ private void recoverPrimary(Set<ServerLocation> excludedServers) {
           excludedServers.add(newPrimary.getServer());
           newPrimary = null;
         }
+
+        markAsQueueAsReadyForEvents(newPrimary);
```

Comment:
The "readyForEvents" op, along with indicating client is ready  and starting the dispatcher thread on the server; it also adds a Marker message; which dictates whether the events delivered from server is applied to local client cache. Any events that arrives before the marker message is not applied to local cache.

The sequence of action that should happen is:
- Get the primary connection
- Clear the local cache
- Send readyForEvent message (it adds the marker message behind the events in server side queue)
- Do interest registration (this gets those events before the marker as part of the response, so need to apply to local cache)

You need to have a test; where data is getting added to server cache; while readyForEvents and register-interest in progress; and then compare the entries in cliient and server cache/region.

Reply:
I wanted to see if I broke anything, I was not looking for feedback. I am not done with this bug. Thank you for the feedback though.

Reply:
I got the impression from the bug, that the ready for events should be after the register interest. I know I have seen us previously fix a test like that back in the days of the green team. As far as the clear of the local cache, I am still trying to figure out where that gets triggered.

Reply:
So I talked with Jinmei and I am satisfied that I have the correct ordering based on her understanding and mine. As far as a test goes, there is a test that is already in place, but it only fails periodically on my machine and not in the cloud. I will update that test with a couple of very minor changes, but I think that it (the test) is pretty much ready to go. There are more tests to run though beyond the one that reveals this issue.